### PR TITLE
Add features to distributed Adam for Megatron support

### PR DESCRIPTION
**Pull request by timmoon10**
This PR adds the following features to `DistributedFusedAdam`:
- Gradient clipping with a sub-set of parameters (for model-parallel parameters)
- Support for [`torch.cuda.amp.GradScaler`](https://pytorch.org/docs/stable/amp.html#torch.cuda.amp.GradScaler)
- Support for models with multiple dtypes and devices
- Option to accumulate gradients in buckets (for accumulating FP16 grads in FP32 buffers)

### CODE REVIEW

File: apex/contrib/optimizers/distributed_fused_adam.py [hunk 1]
```diff
@@ -648,95 +647,152 @@ def grad_sync(self):
...
+                p-norm (default: 2)
+
+        """
+        assert max_norm > 0
+        total_norm = self.grad_norm(parameters=parameters, norm_type=norm_type)
+        inv_clip_coef = (total_norm + 1e-6) / max_norm
+        self._inv_grad_scale = torch.clamp(inv_clip_coef, min=1.0).view(1)
+        return total_norm
+
+    def step(self, grad_scaler=None):
```
Comment (crcrpar) [inline apex/contrib/optimizers/distributed_fused_adam.py:764]: My gut feeling says it's not safe to remove `closure`

File: apex/contrib/optimizers/distributed_fused_adam.py [hunk 2]
```diff
@@ -648,95 +647,152 @@ def grad_sync(self):
...
+        # Note: We compute gradient norm to check for non-finite
+        # values. This is more conservative and compute intensive than
+        # directly checking, but it avoids extra communication if we
+        # have already computed gradient norm e.g. for gradient
+        # clipping.
+        if grad_scaler is not None:
+            grad_norm = self.grad_norm()
+            found_inf = torch.logical_not(torch.isfinite(grad_norm))
+            scaler_state = grad_scaler._per_optimizer_states[id(self)]
+            scaler_state['found_inf_per_device'] = {found_inf.device: found_inf.float()}
```
Comment (crcrpar) [inline apex/contrib/optimizers/distributed_fused_adam.py:786]: This cast here is redundant if you assume that `grad_scaler` is an instance of `torch.cuda.amp.GradScaler`
```suggestion
            scaler_state['found_inf_per_device'] = {found_inf.device: found_inf}
```

File: apex/contrib/optimizers/distributed_fused_adam.py [hunk 3]
```diff
@@ -648,95 +647,152 @@ def grad_sync(self):
...
+            if found_inf.item():
+                return
+            else:
+                assert grad_scaler._scale is not None
+                self._inv_grad_scale *= grad_scaler._scale
+        inv_grad_scale = self._inv_grad_scale.item()
```
Comment (crcrpar) [inline apex/contrib/optimizers/distributed_fused_adam.py:792]: This could incur a device-to-host copy.

File: apex/contrib/optimizers/distributed_fused_adam.py (hunk 1 above)
Comment (crcrpar) [inline apex/contrib/optimizers/distributed_fused_adam.py:764]: also, you can define `grad_scaler` as a keyword only argument given https://github.com/pytorch/pytorch/blob/fe73528a904aa0627b5a6f7376a2331cb5985bfa/torch/cuda/amp/grad_scaler.py#L329.

Comment (timmoon10) [reply apex/contrib/optimizers/distributed_fused_adam.py]: Good call, I ran into a problem in PyTorch Lightning without it.

File: apex/contrib/optimizers/distributed_fused_adam.py (hunk 3 above)
Comment (timmoon10) [reply apex/contrib/optimizers/distributed_fused_adam.py:799]: Unfortunately it is needed as a CPU arg in the Adam kernel. I think the next step will be to tweak some of the kernels, but I'd like to keep this PR Python-only for now.

File: apex/contrib/optimizers/distributed_fused_adam.py (hunk 2 above)
Comment (timmoon10) [reply apex/contrib/optimizers/distributed_fused_adam.py:793]: I don't think `GradScaler` does any casting. I got runtime errors when I passed in a `bool`.

Comment (crcrpar) [reply apex/contrib/optimizers/distributed_fused_adam.py:793]: uh this logic overloads `torch.cuda.amp.GradScaler`'s `found_inf` logic. I see

File: apex/contrib/optimizers/distributed_fused_adam.py (hunk 3 above)
Comment (crcrpar) [reply apex/contrib/optimizers/distributed_fused_adam.py:799]: makes sense.

**Review by crcrpar — APPROVED** [review]
Verified the unittest with nightly, `nvcr.io/nvidia/pytorch:22.06-py3`, `pytorch/pytorch:1.12.0-cuda11.3-cudnn8-devel`
